### PR TITLE
Google & Microsoft calendar API integration

### PR DESCRIPTION
**Pull request by damencho**

### CODE REVIEW

File: react/features/calendar-sync/functions.web.js [hunk 1]
```diff
@@ -65,14 +70,25 @@ export function _fetchCalendarEntries(
...
+            break;
+        case CALENDAR_TYPE.MICROSOFT:
+            calendarIntegrationInstance
+                = new MicrosoftCalendarApi(
+                    config.microsoftApiApplicationClientID, store);
+            break;
+        }
     }
+
+    return calendarIntegrationInstance;
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/functions.web.js]: I'm unsure how I feel about the calendarIntegration Instance being stored in this file. I think it's fine but I also feel it might be something that could be stored in the redux state instead or held onto by some singleton.

File: react/features/calendar-sync/middleware.js [hunk 2]
```diff
@@ -79,6 +80,11 @@ StateListenerRegistry.register(
     /* listener */ (googleAPIState, { dispatch }) => {
         dispatch(setCalendarAPIState(googleAPIState));
     });
+StateListenerRegistry.register(
+    /* selector */ state => state['features/google-api'].profileEmail,
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/middleware.js]: It feels off to me that there is now duplicated state. I think you mentioned you weren't sure about the whole profileEmail thing. Why does the same state need to be in calendar and google api? Could the components needing the email look in multiple places for it instead, maybe through a selector/helper that knows where to look in redux?

Or maybe they're separate enough concerns to keep the two separate state entries?

File: react/features/calendar-sync/web/googleCalendar.js [hunk 3]
```diff
@@ -1,37 +1,47 @@
...
      *
-     * @param {Dispatch} dispatch - The redux {@code dispatch} function.
-     * @param {Function} getState - The redux function that gets/retrieves the
-     * redux state.
      * @returns {function(Dispatch<*>): Promise<void>}
      */
-    init(dispatch: Function, getState: Function): Promise<void> {
-        if (getState()['features/calendar-sync'].apiState !== 0) {
+    init(): Promise<void> {
+        if (this._getState()['features/calendar-sync'].apiState !== 0) {
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/googleCalendar.js]: btw, the 0 should probably be the constant for the google api state.

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 4]
```diff
@@ -1,7 +1,492 @@
...
+     * @param {number} fetchEndDays - The number of days to fetch.
+     * @returns {function(Dispatch<*>): Promise<CalendarEntries>}
+     */
+    getCalendarEntries(
+            fetchStartDays: ?number, fetchEndDays: ?number) {
+        return (dispatch: Dispatch<*>, getState: Function): Promise<*> => {
+
+            // not authorized, skip
+            if (!getState()['features/calendar-sync'].msAuthState) {
+                return Promise.resolve();
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: Would a reject be better here?

File: package.json [hunk 5]
```diff
@@ -47,9 +47,11 @@
     "jquery-contextmenu": "2.4.5",
     "jquery-i18next": "1.2.0",
     "js-md5": "0.6.1",
+    "jsrsasign": "8.0.12",
     "jwt-decode": "2.2.0",
     "lib-jitsi-meet": "github:jitsi/lib-jitsi-meet#2be752fc88ff71e454c6b9178b21a33b59c53f41",
     "lodash": "4.17.4",
+    "@microsoft/microsoft-graph-client": "1.1.0",
```
[the remark this answers is not in the document]
Comment (damencho) [reply package.json]: Yes, manually. I see. I will move it.

File: react/features/google-api/constants.js [hunk 6]
```diff
@@ -1,14 +1,23 @@
...
 /**
- * The Google API scopes to request access to for streaming.
+ * The Google API scopes to request access for streaming and calendar.
  *
  * @type {Array<string>}
  */
 export const GOOGLE_API_SCOPES = [
-    'https://www.googleapis.com/auth/youtube.readonly'
+    'https://www.googleapis.com/auth/youtube.readonly',
+    'https://www.googleapis.com/auth/calendar'
```
Comment (virtuacoplenny) [inline react/features/google-api/constants.js]: Are there any changes needed to deployed applications to get calendar permissions?

File: react/features/google-api/googleApi.js [hunk 7]
```diff
@@ -223,6 +225,79 @@ const googleApi = {
...
+                    = calendarList.result.items.map(en => en.id);
+                const promises = calendarIds.map(id => {
+                    const startDate = new Date();
+                    const endDate = new Date();
+
+                    startDate.setDate(startDate.getDate() + fetchStartDays);
+                    endDate.setDate(endDate.getDate() + fetchEndDays);
+
+                    return this._getGoogleApiClient()
+                        .client.calendar.events.list({
```
Comment (virtuacoplenny) [inline react/features/google-api/googleApi.js]: Did you look at the google api request batching? I'm wondering if it's super easy.

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+                const client = Client.init({
+                    authProvider: done => done(
+                        null,
+                        getState()['features/calendar-sync']
+                            .msAuthState.accessToken)
+                });
+
+                const query = client.api('/me/calendars');
+
+                query.get((error, response) => {
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: I read this in the microsoft-graph-client readme: `The actions(.get(), .put(), etc.) accept a callback or don't pass in a function to get back a Promise.` That might mean you don't need the wrapper.

File: react/features/calendar-sync/constants.js [hunk 9]
```diff
@@ -11,27 +11,62 @@ import { NativeModules } from 'react-native';
...
+
+/**
+ * The number of days to go back when fetching.
+ */
+export const FETCH_START_DAYS = -1;
+
+/**
+ * The max number of events to fetch from the calendar.
+ */
+export const MAX_LIST_LENGTH = 10;
```
[the remark this answers is not in the document]
Comment (virtuacoplenny) [reply react/features/calendar-sync/constants.js]: Nevermind, these were always separated. I'd prefer not changing existing code as much as possible for now. Any comment requesting change to existing logic you should just say no.

File: static/msredirect.html [hunk 10]
```diff
@@ -0,0 +1,11 @@
+<html>
```
Comment (virtuacoplenny) [inline static/msredirect.html]: I'm wondering if it's best to include the doctype. According to W3C it is required but I don't think omitting it actually causes much harm for this case.

File: react/features/calendar-sync/functions.any.js [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+
+                if (url) {
+                    return url.toString();
+                }
+            }
+        }
+    }
+
+    return null;
+}
```
[the remark this answers is not in the document]
Comment (damencho) [reply react/features/calendar-sync/functions.any.js]: Yes, there was only one change I made. https://github.com/jitsi/jitsi-meet/pull/3340/commits/7f8a1e2c1799e6bb78a3445e2c3db2f88f6d91fe#diff-e626ccf6b0437daab4a769da852e453dR24
It was executing one unnecessary fetch every time because of comparing arrays, at least that what I saw on mobile, not sure whether that array === array is working differently on mobile.

File: react/features/calendar-sync/functions.web.js [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+let calendarIntegrationInstance;
+
+/**
+ * Determines whether the calendar feature is enabled by the web.
+ *
+ * @returns {boolean} If the app has enabled the calendar feature, {@code true};
+ * otherwise, {@code false}.
+ */
+export function _isCalendarEnabled() {
+    return config.enableCalendarIntegration === true;
```
[the remark this answers is not in the document]
Comment (damencho) [reply react/features/calendar-sync/functions.web.js]: I was not sure how to do that in the constants, as this was the way it was implemented for mobile, so I didn't want to change that logic.

File: react/features/calendar-sync/functions.any.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (virtuacoplenny) [reply react/features/calendar-sync/functions.any.js]: I see. So instead of comparing array values it was comparing if they're pointers to the same array.

File: react/features/calendar-sync/functions.web.js [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+            break;
+        case CALENDAR_TYPE.MICROSOFT:
+            calendarIntegrationInstance
+                = new MicrosoftCalendarApi(
+                    config.microsoftApiApplicationClientID, store);
+            break;
+        }
+    }
+
+    return calendarIntegrationInstance;
```
[the remark this answers is not in the document]
Comment (damencho) [reply react/features/calendar-sync/functions.web.js]: I don't know really... The idea is going with google or microsoft integration and that's it. There is no way to delete that at the moment, maybe we will need something when we implement disconnect calendar button ...

File: react/features/calendar-sync/middleware.js [hunk 14]
```diff
@@ -85,120 +75,16 @@ CALENDAR_ENABLED
...
-            .catch(reject);
+StateListenerRegistry.register(
+    /* selector */ state => state['features/google-api'].googleAPIState,
+    /* listener */ (googleAPIState, { dispatch }) => {
+        dispatch(setCalendarAPIState(googleAPIState));
+    });
+StateListenerRegistry.register(
+    /* selector */ state => state['features/google-api'].profileEmail,
+    /* listener */ (profileEmail, { dispatch }) => {
+        dispatch(setCalendarProfileEmail(profileEmail));
```
[the remark this answers is not in the document]
Comment (damencho) [reply react/features/calendar-sync/middleware.js]: So the youtube stream dialog uses directly the google-api state and that state, and the idea is here that calendar-sync will use google-api and its state, or the Microsoft implementation which will provide its own state for profileEmail.

File: react/features/calendar-sync/web/googleCalendar.js [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+     *
+     * @returns {function(Dispatch<*>): Promise<void>}
+     */
+    init(): Promise<void> {
+        if (this._getState()['features/calendar-sync'].apiState !== 0) {
+            return Promise.resolve();
+        }
+
+        return this._dispatch(loadGoogleAPI(this._appClientID));
+    }
```
[the remark this answers is not in the document]
Comment (damencho) [reply react/features/calendar-sync/web/googleCalendar.js]: Yeah initializing the api ... maybe we can rename it...

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 16]
```diff
@@ -0,0 +1,492 @@
...
+                    } else {
+                        const calendarIds
+                            = response.value.map(en => en.id);
+                        const promises
+                            = calendarIds.map(id =>
+                                this._getCalendarEntries(
+                                    client, id, fetchStartDays, fetchEndDays));
+
+                        return Promise.all(promises)
+                            .then(result => [].concat(...result))
```
[the remark this answers is not in the document]
Comment (damencho) [reply react/features/calendar-sync/web/microsoftCalendar.js]: To flatten it, cause it is an array of arrays of events, I just need one array of events.

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 17]
```diff
@@ -0,0 +1,492 @@
...
+     */
+    _getCalendarEntries(
+            client,
+            calendarId,
+            fetchStartDays: ?number,
+            fetchEndDays: ?number): Promise<*> {
+        /* eslint-enable max-params */
+
+        if (this._isSignedIn()) {
+            this._dispatch(setCalendarAPIState(CALENDAR_API_STATES.SIGNED_IN));
```
[the remark this answers is not in the document]
Comment (damencho) [reply react/features/calendar-sync/web/microsoftCalendar.js]: If you have token that is not expired and you reload the page, this will set the correct state of the calendar-sync that it is signed in, have a correct token ... Maybe this state should be done in one place when we init the api or when we construct the MicrosoftCalendarApi object ..., WDYT?

The fact is that currently when you load the welcome page it calls _fetchCalendarEntries which end up calling getCalendarEntries and I was using that to set the correct state of the feature ...

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 18]
```diff
@@ -0,0 +1,492 @@
...
+            return null;
+        }
+
+        this._dispatch(setCalendarAPIAuthState({
+            authState: undefined,
+            accessToken: tokenInfo.access_token,
+            tokenExpires: expires.getTime(),
+            idToken
+        }));
+        this._dispatch(setCalendarProfileEmail(payload.preferred_username));
```
[the remark this answers is not in the document]
Comment (damencho) [reply react/features/calendar-sync/web/microsoftCalendar.js]: Ok, I will change it, thanks. The caller already dispatches signed_id state ... so it should also dispatches and the rest ....

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 19]
```diff
@@ -0,0 +1,492 @@
+/* @flow */
```
[the remark this answers is not in the document]
Comment (damencho) [reply react/features/calendar-sync/web/microsoftCalendar.js]: Nope, this is not handled yet.

File: react/features/google-api/constants.js (hunk 6 above)
Comment (damencho) [reply react/features/google-api/constants.js]: I think so, you need to just enable that API. We should take care of that about meet.jit.si. I was also thinking of dropping a simple README with links to Google and Microsoft dashboard how to enable these.

File: react/features/google-api/googleApi.js (hunk 7 above)
Comment (damencho) [reply react/features/google-api/googleApi.js]: You mean to get rid of some of the Promises part?

File: react/features/calendar-sync/actions.js [hunk 20]
```diff
@@ -61,3 +66,112 @@ export function setCalendarEvents(events: Array<Object>) {
...
+    return (dispatch: Dispatch<*>, getState: Function): Promise<*> => {
+        const { calendarType } = getState()['features/calendar-sync'];
+        const api = _getCalendarIntegration(
+            calendarType, {
+                dispatch,
+                getState
+            });
+
+        if (!api) {
+            return Promise.reject('No calendar type selected!');
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/actions.js]: Would it make sense to do this check even before trying to get the api?

Comment (virtuacoplenny) [reply react/features/calendar-sync/actions.js]: Nevermind, I get what you're doing. In case some unexpected type is set you want to error here.

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 21]
```diff
@@ -0,0 +1,531 @@
...
+     * @param {Object} store - The redux store.
+     */
+    constructor(appClientID: string, store: Object) {
+        this._appClientID = appClientID;
+        this._dispatch = store.dispatch;
+        this._getState = store.getState;
+        this._currentRequest = {};
+
+        // checks post messages for events from the popup
+        window.addEventListener('message', this._onPostMessage.bind(this));
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: If switching between microsoft and google, would this listener cause any problems? Do you need to do any cleanup in a deconstructor/destroy method?

Comment (damencho) [reply react/features/calendar-sync/web/microsoftCalendar.js]: I don't see how it can cause any problems, but cleaning is something we should take care. I will check it now.

File: react/features/calendar-sync/web/googleCalendar.js [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+     * @returns {function(Dispatch<*>): Promise<CalendarEntries>}
+     */
+    getCalendarEntries,
+
+    /**
+     * Returns the type of calendar integration this object implements.
+     *
+     * @returns {string}
+     */
+    getType() {
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/googleCalendar.js]: I don't think this function is needed anymore.

File: react/features/calendar-sync/web/googleCalendar.js [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+} from '../../google-api';
+import { CALENDAR_TYPE } from '../constants';
+
+declare var config: Object;
+
+/**
+ * A stateless object that implements the expected interface for interacting
+ * Google authentication in order to get calendar data.
+ *
+ * @type {Object}
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/googleCalendar.js]: Probably the doc should mention how pretty much this object is an aggregation of actions that need to be dispatched.

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 24]
```diff
@@ -0,0 +1,564 @@
...
+     *
+     * @type {string}
+     */
+    MS_API_SCOPES: 'openid profile Calendars.Read',
+
+    /**
+     * See https://docs.microsoft.com/en-us/azure/active-directory/develop/
+     * v2-oauth2-implicit-grant-flow#send-the-sign-in-request.
+     *
+     * FIXME BEFORE MERGE: Why is this needed?
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: Need to look into this.

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 25]
```diff
@@ -0,0 +1,564 @@
...
+/**
+ * Store the window from an auth request. That way it can be reused if a new
+ * request comes in and it can be used to indicate a request is in progress.
+ *
+ * @private
+ * @type {Object|null}
+ */
+let popupAuthWindow = null;
+
+export const microsoftCalendarApi = {
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: Should add a jsdoc here.

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 26]
```diff
@@ -0,0 +1,564 @@
...
+     *
+     * @param {number} fetchStartDays - The number of days to go back
+     * when fetching.
+     * @param {number} fetchEndDays - The number of days to fetch.
+     * @returns {function(Dispatch<*>): Promise<CalendarEntries>}
+     */
+    getCalendarEntries(fetchStartDays: ?number, fetchEndDays: ?number) {
+        return (dispatch: Dispatch<*>, getState: Function): Promise<*> => {
+            const state = getState()['features/calendar-sync'] || {};
+            const token = state.msAuthState && state.msAuthState.accessToken;
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: Some token validation function should be used here maybe. Maybe it's in the token validation function that the token refreshing should happen.

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 27]
```diff
@@ -0,0 +1,564 @@
...
+    },
+
+    /**
+     * Returns the type of calendar integration this object implements.
+     *
+     * @returns {string}
+     */
+    getType(): string {
+        return CALENDAR_TYPE.MICROSOFT;
+    },
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: Not needed anymore.

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 28]
```diff
@@ -0,0 +1,564 @@
...
+    signIn(): Function {
+        return (dispatch: Dispatch<*>, getState: Function) => {
+            // Ensure only one popup window at a time.
+            if (popupAuthWindow) {
+                popupAuthWindow.focus();
+
+                return Promise.reject('Sign in already in progress.');
+            }
+
+            // FIXME BEFORE MERGE: handle case of user closing the modal.
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: This fixme can be removed now. It is handled with setInterval below.

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 29]
```diff
@@ -0,0 +1,564 @@
...
+
+    /**
+     * Sign out from the Microsoft API Client Library.
+     *
+     * @returns {function(Dispatch<*>): Promise<void>}
+     */
+    signOut(): Function {
+        // Using the Outlook API Get Started guide, there is no revoking to be
+        // done with the Microsoft API. Instead any stored state on the app-side
+        // should be cleared and the token should be allowed to expire.
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: Maybe add a link to the Outlook API guide. Maybe.

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 30]
```diff
@@ -0,0 +1,564 @@
...
+     *
+     * @returns {function(Dispatch<*>): Promise<void>}
+     */
+    updateProfile(): Function {
+        return (dispatch: Dispatch<*>, getState: Function) => {
+            const { msAuthState = {} }
+                = getState()['features/calendar-sync'] || {};
+            const email = msAuthState.userSigninName || '';
+
+            dispatch(setCalendarProfileEmail(email));
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: This is kind of odd. I wonder if it's better to just make an extra call here to get the email or do more in the reducers. It just feels off to me that this action is essentially looking at state elsewhere in redux and duplicating it.

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 31]
```diff
@@ -0,0 +1,564 @@
...
+            iframe.setAttribute('style', 'display: none');
+            iframe.setAttribute('src', refreshAuthUrl);
+
+            const signInPromise = new Promise(resolve => {
+                iframe.onload = () => {
+                    resolve(iframe.contentWindow.location.hash);
+                };
+            });
+
+            if (!document.body) {
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: why did I add this?

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 32]
```diff
@@ -0,0 +1,564 @@
...
+    // to account for differences in clock settings and convert to ms.
+    const expiresIn = (parseInt(params.get('expires_in'), 10) - 300) * 1000;
+    const now = new Date();
+    const expireDate = new Date(now.getTime() + expiresIn);
+
+    params.set('tokenExpires', expireDate.getTime().toString());
+
+    // FIXME: what does this do?
+    return Array.from(params.entries())
+        .reduce((main, [ key, value ]) => {
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: "main" is the accumulator? Why convert this to an object?

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 33]
```diff
@@ -0,0 +1,564 @@
...
+ *
+ * @param {Object} client - The Microsoft-graph-client initialized.
+ * @param {string} calendarId - The calendar ID to use.
+ * @param {number} fetchStartDays - The number of days to go back
+ * when fetching.
+ * @param {number} fetchEndDays - The number of days to fetch.
+ * @returns {Promise<any> | Promise}
+ * @private
+ */
+function requestCalendarEvents(
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: the eslint disabling can be done on this line, I think, and using "eslint-disable-line" instead of "eslint-disable" plus "eslint-enable".

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 34]
```diff
@@ -0,0 +1,564 @@
...
+        .select('id,subject,start,end,location,body')
+        .orderby('createdDateTime DESC')
+        .get();
+}
+
+/**
+ * Something.
+ *
+ * FIXME BEFORE MERGE: Understand what exactly these Microsoft-provided
+ * functions do.
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: Need to do this.

File: react/features/google-api/actions.js [hunk 35]
```diff
@@ -14,9 +14,16 @@ import googleApi from './googleApi';
  * @returns {Function}
  */
 export function loadGoogleAPI(clientId: string) {
-    return (dispatch: Dispatch<*>) =>
-        googleApi.get()
-        .then(() => googleApi.initializeClient(clientId))
+    return (dispatch: Dispatch<*>, getState: Function) =>
+        googleApi.load()
```
Comment (virtuacoplenny) [inline react/features/google-api/actions.js]: I think this can still be .get().

File: react/features/google-api/googleApi.js (hunk 7 above)
Comment (virtuacoplenny) [reply react/features/google-api/googleApi.js]: No, to make it so there aren't multiple requests going out at once and instead are batched into one request. Now I think we should punt on it considering the breadth of this PR.

File: react/features/google-api/googleApi.js [hunk 36]
```diff
@@ -223,6 +239,85 @@ const googleApi = {
...
+                            'timeMin': startDate.toISOString(),
+                            'timeMax': endDate.toISOString(),
+                            'showDeleted': false,
+                            'singleEvents': true,
+                            'orderBy': 'startTime'
+                        });
+                });
+
+                return Promise.all(promises)
+                    .then(result =>
```
Comment (virtuacoplenny) [inline react/features/google-api/googleApi.js]: I think naming this "results" might make it more obvious, if it isn't already, that there are multiple results in an array that need to be flattened.

File: react/features/settings/components/web/CalendarTab.js [hunk 37]
```diff
@@ -0,0 +1,286 @@
...
+
+    /**
+     * Whether or not to displaying signing in to Microsoft.
+     */
+    _enableMicrosoftIntegration: boolean,
+
+    /**
+     * The current calendar integration in use, if any.
+     */
+    _isConnectedToCalendar: ?Object,
```
Comment (virtuacoplenny) [inline react/features/settings/components/web/CalendarTab.js]: this is a boolean now.

File: react/features/settings/components/web/CalendarTab.js [hunk 38]
```diff
@@ -0,0 +1,286 @@
...
+
+    /**
+     * Loads third party APIs as needed and bootstraps the initial calendar
+     * state if not already set.
+     *
+     * @inheritdoc
+     */
+    componentDidMount() {
+        this.props.dispatch(bootstrapCalendarIntegration())
+            .catch(() => console.warn('this error right here...'))
```
Comment (virtuacoplenny) [inline react/features/settings/components/web/CalendarTab.js]: Gotta remove this debugging code.

File: react/features/settings/components/web/CalendarTab.js [hunk 39]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+/**
+ * Maps (parts of) the Redux state to the associated props for the
+ * {@code CalendarTab} component.
+ *
+ * @param {Object} state - The Redux state.
+ * @private
+ * @returns {{
+ *     _enableGoogleIntegration: boolean,
```
Comment (virtuacoplenny) [inline react/features/settings/components/web/CalendarTab.js]: Have to update this jsdoc for the proper return.

File: react/features/settings/components/web/SettingsDialog.js [hunk 40]
```diff
@@ -1,19 +1,23 @@
...
 } from '../../../device-selection';
 
+import CalendarTab from './CalendarTab';
 import MoreTab from './MoreTab';
 import ProfileTab from './ProfileTab';
-import { getMoreTabProps, getProfileTabProps } from '../../functions';
+import {
+    getMoreTabProps,
+    getProfileTabProps
+} from '../../functions';
```
Comment (virtuacoplenny) [inline react/features/settings/components/web/SettingsDialog.js]: Undo this change.

File: react/features/calendar-sync/web/microsoftCalendar.js (hunk 34 above)
Comment (virtuacoplenny) [reply react/features/calendar-sync/web/microsoftCalendar.js]: It's obvious, really, if one carefully reads the *code* (but not the docs)...

File: react/features/calendar-sync/web/microsoftCalendar.js [hunk 41]
```diff
@@ -0,0 +1,559 @@
...
+
+/**
+ * Converts a hash parameter string into an object.
+ *
+ * @param {string} hash - The string to parse.
+ * @private
+ * @returns {Object}
+ */
+function getParamsFromHash(hash) {
+    const params = new URLSearchParams(hash.slice(1, hash.length));
```
Comment (virtuacoplenny) [inline react/features/calendar-sync/web/microsoftCalendar.js]: Looks like "parseURLParams" exists somewhere and we should use that.

File: react/features/calendar-sync/functions.web.js [hunk 42]
```diff
@@ -0,0 +1,93 @@
...
+    const { integrationType } = getState()['features/calendar-sync'];
+    const integration = _getCalendarIntegration(integrationType);
+
+    if (!integration) {
+        logger.debug('No calendar type available');
+
+        return;
+    }
+
+    dispatch(integration.load())
```
[the remark this answers is not in the document]
Comment (virtuacoplenny) [reply react/features/calendar-sync/functions.web.js]: ignoring for now. this is how it has been.

File: react/features/calendar-sync/functions.web.js [hunk 43]
```diff
@@ -0,0 +1,93 @@
...
+        .then(() => dispatch(integration._isSignedIn()))
+        .then(signedIn => {
+            if (signedIn) {
+                return Promise.resolve();
+            }
+
+            return Promise.reject('Not authorized, please sign in!');
+        })
+        .then(() => dispatch(integration.getCalendarEntries(
+            FETCH_START_DAYS, FETCH_END_DAYS)))
```
[the remark this answers is not in the document]
Comment (virtuacoplenny) [reply react/features/calendar-sync/functions.web.js]: _isSignedIn also refreshes tokens.
